From 517f8a10a478d7916b37abd3c0fd19c24cecbc2b Mon Sep 17 00:00:00 2001
From: Jens Rottmann <Jens.Rottmann@ADLINKtech.com>
Date: Thu, 2 Apr 2015 18:39:47 +0200
Subject: Fix heavy flickering when moving camera window.

This is done by adding a shortcut code path to the driver: if already
initialized and the overlay window's size is unchanged then just adjust the
position. Otherwise the full setup is done as before.

diff --git a/drivers/media/platform/mxc/capture/ipu_fg_overlay_sdc.c b/drivers/media/platform/mxc/capture/ipu_fg_overlay_sdc.c
index 8cad596..a976c2a 100644
--- a/drivers/media/platform/mxc/capture/ipu_fg_overlay_sdc.c
+++ b/drivers/media/platform/mxc/capture/ipu_fg_overlay_sdc.c
@@ -374,6 +374,15 @@ static int foreground_start(void *private)
 		return -EPERM;
 	}
 
+	/* Shortcut, eliminates heavy flickering when someone simply moves the
+	 * window. If resizing we still do the full setup. */
+	if (cam->overlay_active && fbi->var.xres == cam->win.w.width
+	    && fbi->var.yres == cam->win.w.height) {
+		ipu_disp_set_window_pos(disp_ipu, MEM_FG_SYNC, cam->win.w.left,
+		                        cam->win.w.top);
+		return 0;
+	}
+
 	fbvar = fbi->var;
 
 	/* Store the overlay frame buffer's original std */
-- 
2.1.4

